### ISSUE TITLE
AssertionError in Row.__init_loop when using Table wrapper with indexing and column type or precision mismatch

### ISSUE THREAD

**Issue by kaukrise** [PyTables/PyTables#477]
Iterating over rows fails in 3.2.0 in classes wrapping Table, but only under certain conditions, as far as I can tell:

* The columns have different types or numeric precision, e.g. Int32Col and Int16Col
* One column is indexed

Here is a minimal example that fails with an AssertionError:

```python
import tables as t

class MinTable(t.Table):
    
    def __init__(self):
        
        f = t.open_file('bla', 'a', driver="H5FD_CORE",driver_core_backing_store=0)
        
        description = {
            'c1': t.Int32Col(),
            'c2': t.Int16Col()
        }
        
        t.Table.__init__(self, f.get_node('/'), 'test', description=description)
        
        self._enable_index()
        
        row = self.row
        row['c1'] = 0
        row.append()
        self.flush()
        
    def _enable_index(self):
        self.cols.c1.create_index()

m = MinTable()
# next line fails
[x.fetch_all_fields() for x in m.where('c1 == 0')]
```

The error produced is:

```
/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/tables/table.pyc in where(self, condition, condvars, start, stop, step)
   1499         """
   1500
-> 1501         return self._where(condition, condvars, start, stop, step)
   1502
   1503     def _where(self, condition, condvars, start=None, stop=None, step=None):

/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/tables/table.pyc in _where(self, condition, condvars, start, stop, step)
   1538         if profile:
   1539             show_stats("Exiting table._where", tref)
-> 1540         return row._iter(start, stop, step, chunkmap=chunkmap)
   1541
   1542     def read_where(self, condition, condvars=None, field=None,

tables/tableextension.pyx in tables.tableextension.Row._iter (tables/tableextension.c:9112)()

tables/tableextension.pyx in tables.tableextension.Row._init_loop (tables/tableextension.c:10137)()

AssertionError:
```

Changing Int32Col to Int16Col or vice versa produces no error. Also no problem occurs when switching off indexing. This does not happen in the normal Table class, as far as I can tell. The following works fine:

```python
import tables as t

f = t.open_file('bla2', 'a', driver="H5FD_CORE",driver_core_backing_store=0)
table = t.Table(f.get_node('/'),'test',{ 'c1': t.Int32Col(), 'c2': t.Int16Col() },title='test')
table.row['c1'] = 0
table.row.append()
table.flush()
table.cols.c1.create_index()
[x.fetch_all_fields() for x in table.where('c1 == 0')]
```

**Comment by kaukrise** [comment]
Finally got around to fixing this myself.

Lines [373 to 377](https://github.com/PyTables/PyTables/blob/fb9b917142996e7dce427af6b46e3e64b36d8107/tables/leaf.py#L373-L377) and [385 to 387](https://github.com/PyTables/PyTables/blob/fb9b917142996e7dce427af6b46e3e64b36d8107/tables/leaf.py#L385-L387) in [tables/leaf.py](https://github.com/PyTables/PyTables/blob/fb9b917142996e7dce427af6b46e3e64b36d8107/tables/leaf.py) correct for previously known issues with nrowsinbuf calculation. The fixes specifically check if we are dealing with a Table object. The relevant lines are:

```python
if self.__class__.__name__ == "Table":
```
and
```python
if self.chunkshape is not None and self.__class__.__name__ == "Table":
```

This fix does not take effect for classes wrapping Table. Therefore I have simply added checks for the object's base classes like this:

```python
if (self.__class__.__name__ == "Table" or
    "Table" in [base.__name__ for base in self.__class__.__bases__]):
```
and
```python
if (self.chunkshape is not None and (self.__class__.__name__ == "Table" or
    "Table" in [base.__name__ for base in self.__class__.__bases__])):
```
All tests run fine and the above minimal example works now.

I have submitted a pull request (#489).

**Comment by kaukrise** [comment]
Issue resolved with pull request #489. Thanks for the support!
